source/opt/fold.cpp

```diff
@@ -180,6 +182,53 @@ uint32_t OperateWords(SpvOp opcode,
   }
 }
 
+ir::Instruction* FoldInstructionInternal(
+    ir::Instruction* inst, std::function<uint32_t(uint32_t)> id_map,
+    InstructionBuilder<ir::IRContext::kAnalysisDefUse>* ir_builder,
+    bool* kill_inst) {
+  ir::IRContext* context = inst->context();
+  ir::Instruction* folded_inst = FoldInstructionToConstant(inst, id_map);
+  if (folded_inst != nullptr) {
+    if (*kill_inst) {
+      context->KillInst(inst);
+    }
+    return folded_inst;
+  }
+
+  SpvOp opcode = inst->opcode();
+  analysis::ConstantManager* const_manger = context->get_constant_mgr();
+
+  std::vector<uint32_t> ids;
+  std::vector<const analysis::Constant*> constants;
+  for (uint32_t i = 0; i < inst->NumInOperands(); i++) {
+    const ir::Operand* operand = &inst->GetInOperand(i);
+    if (operand->type != SPV_OPERAND_TYPE_ID) {
+      ids.push_back(0);
+      constants.push_back(nullptr);
+      continue;
+    }
+    ids.push_back(id_map(operand->words[0]));
+    const analysis::Constant* constant =
+        const_manger->FindDeclaredConstant(ids[i]);
+    constants.push_back(constant);
+  }
+
+  static FoldingRules rules;
```

Comment:
No static function variable of class type.

Reply:
What about just have another function that switches on opcode and calls an opcode specific folder? The opcode specific folder can then try all the various rules for that opcode.

Reply:
Think I'll change this to a static pointer, and initialize it with a call to new.  The other options is to move it so it lives somewhere else like the IRContext, but I don't like that because I don't want it used by anybody else.

I have a preference to avoid the switch.  Less code for people to look at if the want to try to understand what is going on.